File: full_case_study_solns/_e_ballbeam/matlab_old/hw8/ballbeamController.m
classdef ballbeamController
    % 
    %    This class inherits other controllers in order to organize multiple controllers.
    %
    %----------------------------
    properties
        zCtrl
        thetaCtrl
        m1
        m2
        g
        length
    end
    %----------------------------
    methods
        %----------------------------
        function self = ballbeamController(P)
            self.m1 = P.m1;
            self.m2 = P.m2;
            self.g = P.g;
            self.length = P.length;
            % Instantiates the PD control objects
            self.zCtrl = PDControl(P.kp_z, P.kd_z, P.theta_max, P.beta, P.Ts);
            self.thetaCtrl = PDControl(P.kp_th, P.kd_th, P.Fmax, P.beta, P.Ts);
        end
        %----------------------------
        function F = u(self, y_r, y)
            % y_r is the referenced input
            % y is the current state
            z_r = y_r;
            z = y(1);
            theta = y(2);
            % the reference angle for theta comes from the outer loop PD control
            theta_r = self.zCtrl.PD(z_r, z, false);
            % the force applied to the cart comes from the inner loop PD control
            F_tilde = self.thetaCtrl.PD(theta_r, theta, false);
            % equilibrium force
            Fe = self.m1*self.g*(z/self.length) + self.m2*self.g/2;
            % total force
            F = F_tilde + Fe;
            % saturate the total force
            F = self.thetaCtrl.saturate(F);
        end
    end
end

File: full_case_study_solns/_e_ballbeam/python_old/hw8/ballbeamController.py
import sys
sys.path.append('..')  # add parent directory
import ballbeamParam as P
import ballbeamParamHW8 as P8
from PDControl import PDControl

class ballbeamController:
    ''' 
        This class inherits other controllers in order to organize multiple controllers.
    '''

    def __init__(self):
        # Instantiates the SS_ctrl object
        self.zCtrl = PDControl(P8.kp_z, P8.kd_z, P8.theta_max, P.beta, P.Ts)
        self.thetaCtrl = PDControl(P8.kp_th, P8.kd_th, P.Fmax, P.beta, P.Ts)

    def u(self, y_r, y):
        # y_r is the referenced input
        # y is the current state
        z_r = y_r[0]
        z = y[0]
        theta = y[1]
        # the reference angle for theta comes from the outer loop PD control
        theta_r = self.zCtrl.PD(z_r, z, flag=False)
        # the force applied to the cart comes from the inner loop PD control
        F_tilde = self.thetaCtrl.PD(theta_r, theta, flag=False)
        # equilibrium force
        Fe = P.m1*P.g*(z/P.length) + P.m2*P.g/2.0
        # total force
        F = F_tilde + Fe
        return [F]







